cwltool/workflow.py

```diff
@@ -436,7 +479,79 @@ def __init__(self, toolpath_object, **kwargs):
         self.steps = [WorkflowStep(step, n, **kwargs) for n, step in enumerate(self.tool.get("steps", []))]
         random.shuffle(self.steps)
 
-        # TODO: statically validate data links instead of doing it at runtime.
+        # statically validate data links instead of doing it at runtime.
+        workflow_inputs = self.tool["inputs"]
+        workflow_outputs = self.tool["outputs"]
+
+        step_inputs = []; step_outputs = []
+        for step in self.steps:
+            step_inputs.extend(step.tool["inputs"])
+            step_outputs.extend(step.tool["outputs"])
+
+        # source parameters: workflow_inputs and step_outputs
+        # sink parameters: step_inputs and workflow_outputs
+
+        # make a dictionary of source parameters, indexed by the "id" field
+        src_parms = workflow_inputs + step_outputs
+        src_dict = {}
+        for parm in src_parms:
+            src_dict[parm["id"]] = parm
+
+        SrcSink = namedtuple("SrcSink", ["src", "sink", "linkMerge"])
+
+        def check_all_types(sinks, sourceField):
```

Comment:
Would you clarify this suggestion? I can make this a function under the class Workflow, but I don't think it makes sense for it to be a method of a Workflow object, like check_all_types(self, sinks, sourceField).